spec/abci/abci++_app_requirements.md

```diff
@@ -188,6 +192,34 @@ Likewise, `ExtendVote` can also be non-deterministic:
 * *w<sup>r</sup><sub>p</sub> = w<sup>r</sup><sub>q</sub> &#8655;
   e<sup>r</sup><sub>p</sub> = e<sup>r</sup><sub>q</sub>*
 
+### Mempool Connection Requirements
+
+Before defining our notation, we need to introduce a requirement for `CheckTx` to always
+yield the same result while *p*'s Application is at a given height.
+
+* Requirement 13 [`CheckTx`, same-height immutability]: For any correct process *p*,
+  any transaction *tx*, and any height *h<sub>p</sub>*,
+  `ResponseCheckTx` returns the same result code while *p* is at *h<sub>p</sub>*.
+
+Requirement 13 allows us to use heights rather than time in our notation.
+Let *CheckTxCode<sub>tx,p,h</sub>* denote the result  code returned by *p*'s Application, via `ResponseCheckTx`,
+to CometBFT's `RequestCheckTx` call occurring while the Application is at height *h*
+and having transaction *tx* as parameter.
+Let predicate *OK(CheckTxCode)* denote whether *CheckTxCode* is `SUCCESS`.
+
+* Requirement 14 [`CheckTx`, eventual non-oscillation]: For any correct process *p*
+  and any transaction *tx*, there exists a height *h<sub>p,stable</sub>*
+  such that *OK(CheckTxCode<sub>tx,p,h<sub>p,stable</sub></sub>) = OK(CheckTxCode<sub>tx,p,h</sub>)*
+  for any height *h > h<sub>p,stable</sub>*.
+
+Requirement 14, local to process *p*, ensures that
+a transaction will eventually stop oscillating between `CheckTx` success and failure
+if it stays in *p's* mempool for long enough.
+Thus, whilst being a requirement local to *p*, it ensures that
```

Comment:
> it ensures that a transaction will leave the mempool of *all* full nodes

We cannot ensure that from Requirement 14. This requirement allows two different processes to non-oscillate at different validity values. One process may stabilize on valid while the other on invalid, and not necessarily at the same time/height. To be able to talk about a global property, I think we cannot avoid talking about transactions committed to the chain, as in [INV4](https://github.com/cometbft/knowledge-base/blob/main/protocols/mempool-overview.md).

Reply:
>  One process may stabilize on valid while the other on invalid, and not necessarily at the same time/height.

Why? Since we are using heights and the application is deterministic, the output at the beginning of a given height should be exactly the same.

The problem I see in those properties is that a statefull validation is dependent on the order at which transactions are verified. In particular, Req. 13 is hard to fulfill given this assumption. The most simpler examples regards duplication: if `tx` is checked twice against the application, for instance because the mempool fails preventing duplication, a stateful application should reject the second "copy" of the transaction.

Reply:
> To be able to talk about a global property, I think we cannot avoid talking about transactions committed to the chain

Agreed.

Reply:
The sentence in the text going from a local property (the requirement) to a global one was to explain that we only need a local property to obtain a global one. Of course as you rightly point out in your comments, the global property does not _trivially_ follow from the local one. You need other things (such as consensus agreement) to _prove_ it, but I left that of of the text... maybe by trying to be concise, I ended up being unclear